Adding APIv4 Support and a few other features

attzonko opened the pull request:
This is our working version of the BOT which integrates a few of the outstanding pull requests as well as adds support for APIv4. Many of the changes including the APIv4 support were previously contributed by @seLain back in Sep'17, I just grabbed them from his repository and got them to work together with a few minor tweaks. Below is a high level summary of the changes. 

- Adding APIv4 Support ([source branch](https://github.com/seLain/mattermost_bot/tree/support_APIv4)) #35
- Enabling local settings  #47 
- Allow posts from multiple teams #46 #45 #39 
- Added missing default plugins #54
- Added ignore 'ignore strings' anywhere in a message #37
- A few other minor cosmetic changes

blampe committed:
Fix an issue where multiple handlers weren't registering correctly for one regex

blampe committed:
Fix typo

blampe committed:
Simplify by using DEBUG to bypass the cache

Commit:
ignore messages with ignore strings anywhere in message, not just the beginning

Commit:
Adding APIv4 Support and a few other features

- Adding APIv4 Support
- Enabling local settings
- Allow posts from multiple teams
- Added missing default plugins

jasonblais commented:
@gotlium let us know if you have time to review this PR. Thanks! :)

attzonko committed:
Merge pull request #1 from attzonko/working_apiv4

Adding APIv4 Support and a few other features

attzonko committed:
Merge pull request #2 from wtodom/ignore-anywhere

ignore messages with ignore strings anywhere in message, not just the…

attzonko committed:
Merge pull request #3 from attzonko/working_apiv4

Pulling the ignore fix from @wtodom

attzonko committed:
Merge pull request #4 from blampe/multiple-plugins

Getting the Multiple plugins Fix

attzonko committed:
Merge pull request #5 from attzonko/working_apiv4

Pulling in the multiple plugin fix

attzonko committed:
Update README.md

Commit:
Minor fixes for APIv4 usage with multiple teams

attzonko committed:
Merge pull request #6 from attzonko/working_apiv4

Minor fixes for APIv4 usage with multiple teams

jasonblais commented:
@gotlium @deniskabalkin Would you have thoughts on the best way to transfer ownership of https://pypi.python.org/pypi/mattermost_bot?

@attzonko is open to assuming ownership of the repo and/or forking it for now.

epatpol commented:
I was actually looking into making the bot work with apiv4, but it seems there's already a PR opened for it (awesome!). So for that I hope @gotlium gives a pong back soon :)

Commit:
Refactoring how user information is aquired and removing dead code

Sebastien-Meiffren commented:
Hi,
I've been trying to use your new version of the bot, but I keep having a team issue
```
Traceback (most recent call last):
  File "C:/Users/chuiv/PycharmProjects/bot_mattermost/run.py", line 5, in <module>
    Bot().run()
  File "C:\Users\chuiv\PycharmProjects\bot_mattermost\mattermost_bot\bot.py", line 29, in __init__
    settings.SSL_VERIFY)
  File "C:\Users\chuiv\PycharmProjects\bot_mattermost\mattermost_bot\mattermost_v4.py", line 110, in __init__
    self.login(team, email, password)
  File "C:\Users\chuiv\PycharmProjects\bot_mattermost\mattermost_bot\mattermost.py", line 184, in login
    self.user = self.api.login(team, email, password)
  File "C:\Users\chuiv\PycharmProjects\bot_mattermost\mattermost_bot\mattermost_v4.py", line 25, in login
    self.load_initial_data()
  File "C:\Users\chuiv\PycharmProjects\bot_mattermost\mattermost_bot\mattermost_v4.py", line 34, in load_initial_data
    self.default_team_id = self.teams[0]['id']
IndexError: list index out of range
```

If I print `self.teams = self.get('/teams')`, I get an empty array `[]`

If I connect to the webclient like a normal person / user, everything works fine

attzonko commented:
@Sebastien-Meiffren I went ahead and opened an issue for you in my fork, for now it would be easier to resolve this there and push the change/fix to this pull request. Here is the [link](https://github.com/attzonko/mattermost_bot/issues/7) to the issue.

seLain committed:
fixed the issue which only open (to join) teams gets loaded at login

this was caused by misuse of APIv4, the '/teams' endpoint returns only open teams. we need to use '/users/{user_id}/teams' to get all teams

referred discussion: https://forum.mattermost.org/t/solved-mattermost-v3-to-v4-upgrade-team-api-v4/4156/6

attzonko committed:
Update codacy link

seLain committed:
add tests

seLain committed:
fix codacy reported issues

seLain committed:
fix codacy reported issues

seLain committed:
remvoe trailing whitespace

epatpol commented:
Sorry to hijack this PR but as a python beginner what's the best way to use your fork meanwhile? I used pip to get the latest version from LPgenerator but I would like to test out these changes at the moment.

seLain commented:
I would do it this way without removing installed mattermost_bot (from PyPI) :
1. create a virtual environment __mmpy_bot_test__
`python -m venv mmpy_bot_test`
2. activate the virtual environment
`mmpy_bot_test\Scripts\activate`
3. install attzonko's fork:master
`pip install git+https://github.com/attzonko/mattermost_bot.git`
4. then you can test your bot under this virtual environment

timball commented:
MM api docs claim that the v3 api is deprecated as of 2018-01-16. 

How far away is this code from being able to be merged?

--timball

attzonko commented:
@timball Not sure the original author is ever planning to merge this in. Feel free to use my fork of this project. I plan on making a new pypi package shortly.

timball commented:
Oh cool. can you update your docs? i tried to follow the example in the
readme and it blows up.

--timball

On Tue, Feb 27, 2018 at 1:34 PM, Alex Tzonkov <notifications@github.com>
wrote:

> @timball <https://github.com/timball> Not sure the original author is
> ever planning to merge this in. Feel free to use my fork of this project. I
> plan on making a new pypi package shortly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LPgenerator/mattermost_bot/pull/55#issuecomment-368979464>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACY8EjTCtDmiih0oJfptZV57oWQQ1nQks5tZEpBgaJpZM4RVLQo>
> .
>



-- 
        GPG key available on sks-keyservers.net
pub   rsa2048/D0A8B0F1638B16EF 2018-01-03 Timothy Lu Hu Ball <
timball@gmail.com>
Primary key fingerprint: 1453 FD65 1D18 A992 621E  E4BE D0A8 B0F1 638B 16EF

seLain committed:
updated to support APIv4 only

- revise test cases
- merge mattermost_v4 to mattermost, remove all APIv3 endpoints
- the existing webhook-related part was revised to APIv4 endpoints, but not fully tested yet.

seLain committed:
update README api v3 -> v4

epatpol commented:
Before you could get around without using `MATTERMOST_BOT_SETTINGS_MODULE` but now you have to? I see that you changed it from os.get to os.environ, any reasons for that?

epatpol commented:
NVM I see that you updated the readme on how to use it thank you :)

timball commented:
actually i figured what i did wrong. typo put `http` instead of `https` and
it gave a non-sensical error. Fixed the typo and i'm off to the races.
thanks!

--timball

On Thu, Mar 1, 2018 at 5:14 PM, Patrick-Jeffrey Pollo Guilbert <
notifications@github.com> wrote:

> NVM I see that you updated the readme on how to use it thank you :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LPgenerator/mattermost_bot/pull/55#issuecomment-369749683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACY8J1uBUCKDS9kMv9b_MF5pk1qt_GYks5taHLHgaJpZM4RVLQo>
> .
>



-- 
        GPG key available on sks-keyservers.net
pub   rsa2048/D0A8B0F1638B16EF 2018-01-03 Timothy Lu Hu Ball <
timball@gmail.com>
Primary key fingerprint: 1453 FD65 1D18 A992 621E  E4BE D0A8 B0F1 638B 16EF

seLain committed:
fix: forget to remove redundant v4 from bot.py

seLain committed:
deal with redirected login request

scottleedavis commented:
@attzonko   I am interested in contributing!  I have a vested interest in seeing this apiV4 support thru and love python too!  :)

attzonko commented:
@scottleedavis Hey we ended up creating a new repo from this instead of a fork to change the name and be able to release pypi packages. take a look at https://github.com/attzonko/mmpy_bot

Would love to have you contribute!
